File: fn/src/main/java/org/sqlbuilder/fn/objects/Sentence.java
package org.sqlbuilder.fn.objects;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.common.Utils;
import org.sqlbuilder.common.HasID;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import edu.berkeley.icsi.framenet.SentenceType;

public class Sentence implements HasID, Insertable
{
	public static final Set<Sentence> SET = new HashSet<>();

	private final int sentenceid;

	private final Integer corpusid;

	private final Integer docid;

	private final int paragno;

	private final int sentno;

	private final int apos;

	private final String text;

	public static Sentence make(final SentenceType sentence)
	{
		var s = new Sentence(sentence);
		Sentence.SET.add(s);
		return s;
	}

	private Sentence(final SentenceType sentence)
	{
		this.sentenceid = sentence.getID();
		this.corpusid = sentence.getCorpID();
		this.docid = sentence.getDocID();
		this.paragno = sentence.getParagNo();
		this.sentno = sentence.getSentNo();
		this.text = sentence.getText();
		this.apos = sentence.getAPos();
	}

	// I D E N T I T Y

	public int getID()
	{
		return sentenceid;
	}

	public Integer getCorpusID()
	{
		return corpusid;
	}

	public Integer getDocID()
	{
		return docid;
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		Sentence sentence = (Sentence) o;
		return sentenceid == sentence.sentenceid && Objects.equals(corpusid, sentence.corpusid) && Objects.equals(docid, sentence.docid);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(sentenceid, corpusid, docid);
	}

	// O R D E R

	public static final Comparator<Sentence> COMPARATOR = Comparator.comparing(Sentence::getID).thenComparing(Sentence::getDocID).thenComparing(Sentence::getCorpusID);

	// I N S E R T

	@Override
	public String dataRow()
	{
		return String.format("%d,%s,%s,%d,%d,'%s',%d", //
				sentenceid, //
				Utils.zeroableInt(corpusid), //
				Utils.zeroableInt(docid), //
				paragno, //
				sentno, //
				Utils.escape(text), //
				apos);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[SENT id=%s text=%s]", sentenceid, ellipsis(text));
	}

	private String ellipsis(String text)
	{
		int max = 32;
		int len = text.length();
		return text.substring(0, Math.min(max, len));
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/objects/Sentence.kt
package org.sqlbuilder.fn.objects

import edu.berkeley.icsi.framenet.SentenceType
import org.sqlbuilder.common.HasID
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.common.Utils.escape
import org.sqlbuilder.common.Utils.zeroableInt
import java.util.*
import kotlin.math.min

class Sentence private constructor(
    sentence: SentenceType,
) : HasID, Insertable {

    val iD: Int = sentence.getID()

    val corpusID: Int = sentence.getCorpID()

    val docID: Int = sentence.getDocID()

    private val paragno: Int = sentence.getParagNo()

    private val sentno: Int = sentence.getSentNo()

    private val apos: Int = sentence.getAPos()

    private val text: String = sentence.getText()

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val sentence = o as Sentence
        return iD == sentence.iD && corpusID == sentence.corpusID && docID == sentence.docID
    }

    override fun hashCode(): Int {
        return Objects.hash(iD, corpusID, docID)
    }

    // I N S E R T

    override fun dataRow(): String {
        return "$iD,${zeroableInt(corpusID)},${zeroableInt(docID)},$paragno,$sentno,'${escape(text)}',$apos"
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[SENT id=$iD text=${ellipsis(text)}]"
    }

    companion object {

        private fun ellipsis(text: String): String {
            val max = 32
            val len = text.length
            return text.substring(0, min(max.toDouble(), len.toDouble()).toInt())
        }

        @JvmField
        val COMPARATOR: Comparator<Sentence> = Comparator
            .comparing<Sentence, Int> { it.iD }
            .thenComparing<Int> { it.docID }
            .thenComparing<Int> { it.corpusID }

        @JvmField
        val SET = HashSet<Sentence>()

        fun make(sentence: SentenceType): Sentence {
            val s = Sentence(sentence)
            SET.add(s)
            return s
        }
    }
}
